tutorial/408_DirectDeltaMush/main.cpp

```diff
@@ -0,0 +1,169 @@
+int p = 20;
+float lambda = 3; // 0 < lambda
+float kappa = 1; // 0 < kappa < lambda
+float alpha = 0.8; // 0 <= alpha < 1
+
+bool pre_draw(igl::opengl::glfw::Viewer & viewer)
+{
+  using namespace Eigen;
+  using namespace std;
+  if (recompute)
+  {
+    // Find pose interval
+    const int begin = (int) floor(anim_t) % poses.size();
+    const int end = (int) (floor(anim_t) + 1) % poses.size();
+    const double t = anim_t - floor(anim_t);
+
+    // Interpolate pose and identity
+    RotationList anim_pose(poses[begin].size());
+    for (int e = 0; e < poses[begin].size(); e++)
+    {
+      anim_pose[e] = poses[begin][e].slerp(t, poses[end][e]);
+    }
+    // Propagate relative rotations via FK to retrieve absolute transformations
+    RotationList vQ;
+    vector<Vector3d> vT;
+    igl::forward_kinematics(C, BE, P, anim_pose, vQ, vT);
+    const int dim = C.cols();
+    MatrixXd T(BE.rows() * (dim + 1), dim);
+    TransformationList T_list;
+    for (int e = 0; e < BE.rows(); e++)
+    {
+      Affine3d a = Affine3d::Identity();
+      a.translate(vT[e]);
+      a.rotate(vQ[e]);
+      T.block(e * (dim + 1), 0, dim + 1, dim) =
+        a.matrix().transpose().block(0, 0, dim + 1, dim);
+      T_list.push_back(a);
+    }
+    if (use_ddm)
+    {
+      igl::direct_delta_mush(V, F, C, BE, T_list, Omega, U);
+    }
+    else
+    {
+      // Compute deformation via LBS as matrix multiplication
+      U = M * T;
+    }
+
+    // Also deform skeleton edges
+    MatrixXd CT;
+    MatrixXi BET;
+    igl::deform_skeleton(C, BE, T, CT, BET);
+
+    viewer.data().set_vertices(U);
+    viewer.data().set_edges(CT, BET, sea_green);
+    viewer.data().compute_normals();
+    if (viewer.core().is_animating)
+    {
+      anim_t += anim_t_dir;
+    }
+    else
+    {
+      recompute = false;
+    }
+  }
+  return false;
+}
+
+bool key_down(igl::opengl::glfw::Viewer & viewer, unsigned char key, int mods)
+{
+  recompute = true;
+  switch (key)
+  {
+    case 'D':
+    case 'd':
+      use_ddm = !use_ddm;
+      return true;
+    case ' ':
+      viewer.core().is_animating = !viewer.core().is_animating;
+      return true;
+    default:
+      return false;
+  }
+}
+
+int main(int argc, char *argv[])
+{
+  using namespace Eigen;
+  using namespace std;
+  igl::readOBJ(TUTORIAL_SHARED_PATH "/arm.obj", V, F);
+  U = V;
+  igl::readTGF(TUTORIAL_SHARED_PATH "/arm.tgf", C, BE);
+  // retrieve parents for forward kinematics
+  igl::directed_edge_parents(BE, P);
+  RotationList rest_pose;
+  igl::directed_edge_orientations(C, BE, rest_pose);
+  poses.resize(4, RotationList(4, Quaterniond::Identity()));
+  const Quaterniond twist(AngleAxisd(igl::PI * 0.8, Vector3d(1, 0, 0)));
+  poses[1][2] = rest_pose[2] * twist * rest_pose[2].conjugate();
+  const Quaterniond bend(AngleAxisd(-igl::PI * 0.7, Vector3d(0, 0, 1)));
+  poses[3][2] = rest_pose[2] * bend * rest_pose[2].conjugate();
+
+  igl::readDMAT(TUTORIAL_SHARED_PATH "/arm-weights.dmat", W);
```

Comment:
For this tutorial, it will be more interesting to compare against piecewise-rigid weights. Here's a hotfix to convert these weights to piecewise-rigid 

```  for(int i = 0; i<W.rows();i++)
  {
    const double Wimax = W.row(i).maxCoeff();
    for(int j = 0; j<W.cols();j++)
    {
      W(i,j) = double( W(i,j) == Wimax);
    }
  }
```
I'm also going to find a more interesting example data that we can read in.